company-posframe.el

```diff
@@ -238,6 +238,21 @@ be triggered manually using `company-posframe-quickhelp-show'."
                         'company-posframe-active-backend-name
                       'company-posframe-inactive-backend-name)))
 
+(defun company-posframe-show-at-point ()
+  "Show `company-posframe' at point."
+  (list :position (point)))
+
+(defun company-posframe-show-at-prefix ()
+  "Show `company-posframe' at prefix."
+  (list :position (- (point) (length company-prefix))
+        :x-pixel-offset (* -1 company-tooltip-margin (default-font-width))))
+
+(defcustom company-posframe-position-handler #'company-posframe-show-at-prefix
+  "Handler for "
```

Comment:
What  "Handler for " mean?

Reply:
This looks weird. `company-posframe` shouldn't have a new custom var.

If anything is required, it should be in company (a new user option; probably a part of the new option we already discussed). This frontend should just use it.

Reply:
> If anything is required, it should be in company (a new user option; probably a part of the new option we already discussed). This frontend should just use it.

ATM the placement logic is in company-posframe - thus I am doing the fix here.

Reply:
The placement functionality is obviously here, but that doesn't mean the information of where to place must be here as well.

Otherwise it would need to be duplicated for every similar backend, current and potential ones.

Reply:
The latest version of the PR is changed, no new variables are introduced, just the structure of the code allows plugging custom poshandler.

Reply:
This looks better.

What I perhaps should have said: it's weird to have the new behavior as optional (instead of just being the new behavior).

But if `company-posframe-show-params` is not supposed to be changed or customized (or, at least, changed very rarely), sounds good.

Reply:
> What I perhaps should have said: it's weird to have the new behavior as optional (instead of just being the new behavior).

I don't get that.

> But if `company-posframe-show-params` is not supposed to be changed or customized (or, at least, changed very rarely), sounds good.

I am planning to use it in lsp-mode reference configuration.

Reply:
Will it become +1 place to customize for the popup to look good?

Reply:
I wouldn't use the word `good` but "consistent with the rest of the functionality of the lsp-mode's reference configuration".

Reply:
Okay, let's phrase it this way. But consistent with what? Some more details would help.

Reply:
I think that I show you the youtube video? The point here is that when the signature popup moves because it is the only option, it makes sense the completion popup to move too. This is what vscode does as well.

Reply:
There are several other things that I would like to make consistent - e. g. signature window to have the same size as the completion window. The company doc popup to be fixed size and with the same width as the other popups like inspecting variables in dap-mode, lsp-ui-doc(which as a side note will be displayed on mouse hover).

Reply:
Ok, now I see you you mean.

Personally, I think VS Code's behavior is inferior, and Company (and Atom as well, it seems https://blog.atom.io/2015/05/15/new-autocomplete.html) perform better-looking positioning. But it's admittedly a matter of taste.

Would it not be possible, though, to align the signature window to the popup window instead, and not vice versa?

Having widths customizable (starting with `company-tooltip-minimum-width` and `company-tooltip-maximum-width`) is fine with me, though ideally there would be as little third-party functionality as possible predicated on particular values to work correctly.

Reply:
> Would it not be possible, though, to align the signature window to the popup window instead, and not vice versa?

I haven't tested that. It might look fine too.

>  though ideally there would be as little third-party functionality as possible predicated on particular values to work correctly.

It will work correctly without all these configurations. The goal of that reference configuration is to start from the optimal result(although a bit opinionated) and let the beginners use it as a base.